### PR TITLE
Stunbaton delayed knockdown now respects antistuns

### PR DESCRIPTION
**Pull request by HMBGERDO**
## What Does This PR Do

This PR moves stunbaton delayed knockdown into status effect
Adds removing this status effect to adrenal bio-chip, changeling adrenals, vampire rejuvenate and to admin rejuvenate
So now if you got batonned and you pressed adrenals to run away you will not fall down in two meters if you didnt get hit after you used it

## Why It's Good For The Game

Thats what adrenals, rejuvenate and other supposed to do. They supposed to reduce or remove all bad effects like stamina damage and other but stunbaton delayed knockdown was ignoring that all, so when you see adrenals guy you can just spam your stunbaton at them and there is high chance that they will not run away

This PR makes them all(adrenals, changeling adrenals, rejuvenate) act like intended. You are great vampire, you just healed tons of damage and all stamina, you SHOULD NOT drop down in two meters cause someone batonned you before

There were other PRs about it bit it seems they died 
https://github.com/ParadiseSS13/Paradise/pull/20715
https://github.com/ParadiseSS13/Paradise/pull/20440

So there is my version of those changings

## Testing

compiled, batonned

## Changelog
:cl:
tweak: If you activate adrenals biochip/epinephrine(changeling)/rejuvenate(vampire)/rejuvenate(admin) after getting batonned, you will not be knockdowned. Works with beepsky, ED209, stunbatons and stunprods
/:cl:

### CODE REVIEW

**Comment by HMBGERDO** [comment]
Moved stunbaton delayed effects to `/datum/status_effect/delayed` which is capable for any delayed action in future, supports preventing action signal and custom duration

File: code/__DEFINES/dcs/signals.dm [hunk 1]
```diff
@@ -861,3 +861,5 @@
 #define COMSIG_OBJECTIVE_CHECK_VALID_TARGET "objective_check_valid_target"
 	#define OBJECTIVE_VALID_TARGET		(1<<0)
 	#define OBJECTIVE_INVALID_TARGET	(1<<1)
+
+#define COMSIG_PREVENT_DELAYED_KNOCKDOWN "prevent_delayed_knockdown"
```
Comment (hal9000PR) [inline code/__DEFINES/dcs/signals.dm:865]: I would call this `COMSIG_LIVING_CLEAR_STUNS`

Comment (HMBGERDO) [reply code/__DEFINES/dcs/signals.dm:865]: but those are not stuns, and we're not clearing em, we're clearing currently delayed knockdown here, are you **_sure_**?

Comment (hal9000PR) [reply code/__DEFINES/dcs/signals.dm:865]: no, you are sending a signal that happens when you clear stuns. 

we are naming a signal not a proc call.

File: code/datums/status_effects/neutral.dm [hunk 2]
```diff
@@ -218,3 +218,32 @@
...
+/datum/status_effect/delayed
+	id = "delayed_status_effect"
+	status_type = STATUS_EFFECT_MULTIPLE
+	alert_type = null
+	var/prevent_signal = null
+	var/datum/callback/expire_proc = null
+
+/datum/status_effect/delayed/on_creation(mob/living/new_owner, new_duration, datum/callback/new_expire_proc, new_prevent_signal = null)
+	if(!duration || !new_expire_proc || !istype(new_expire_proc))
+		qdel(src)
```
Comment (hal9000PR) [inline code/datums/status_effects/neutral.dm:231]: return after this

File: code/datums/status_effects/neutral.dm [hunk 3]
```diff
@@ -218,3 +218,32 @@
...
+	duration = new_duration
+	expire_proc = new_expire_proc
+	. = ..()
+	if(new_prevent_signal)
+		RegisterSignal(owner, new_prevent_signal, PROC_REF(prevent_action))
+		prevent_signal = new_prevent_signal
+
+/datum/status_effect/proc/prevent_action()
+	qdel(src)
```
Comment (hal9000PR) [inline code/datums/status_effects/neutral.dm:240]: use signal handler here

File: code/__DEFINES/dcs/signals.dm [hunk 4]
```diff
@@ -862,4 +862,4 @@
 	#define OBJECTIVE_VALID_TARGET		(1<<0)
 	#define OBJECTIVE_INVALID_TARGET	(1<<1)
 
-#define COMSIG_PREVENT_DELAYED_KNOCKDOWN "prevent_delayed_knockdown"
+#define COMSIG_LIVING_CLEAR_STUNS "prevent_delayed_knockdown"
```
Comment (hal9000PR) [inline code/__DEFINES/dcs/signals.dm:865]: Change the actual string aswell please

File: code/datums/status_effects/neutral.dm [hunk 5]
```diff
@@ -218,3 +218,34 @@
...
+		prevent_signal = new_prevent_signal
+
+/datum/status_effect/proc/prevent_action()
+	SIGNAL_HANDLER
+	qdel(src)
+
+/datum/status_effect/delayed/on_remove()
+	if(prevent_signal)
+		UnregisterSignal(owner, prevent_signal)
+	. = ..()
```
Comment (hal9000PR) [inline code/datums/status_effects/neutral.dm:247]: shouldnt this only remove the signal if there arent any other delayed status effects on them with the same `prevent signal`

Comment (HMBGERDO) [reply code/datums/status_effects/neutral.dm:247]: nah it works well with multiple signals registered with the same datum and signal type

Comment (HMBGERDO) [reply code/datums/status_effects/neutral.dm:247]: well i explored signal system and here is what i got
this unregister signal affects only one datum status effect, which registered signal earlier
once we get new status_effect/delayed, we get new datum with new registered signal stored in them
so unregistering signal on one of them will affect only that one
btw this UnregisterSignal shouldnt do much cause we actually deleting datum in this proc so anyway signal will be gone
if you're worried about this, i can remove this string
image

File: code/datums/status_effects/neutral.dm [hunk 6]
```diff
@@ -218,3 +218,34 @@
...
+
+/datum/status_effect/delayed
+	id = "delayed_status_effect"
+	status_type = STATUS_EFFECT_MULTIPLE
+	alert_type = null
+	var/prevent_signal = null
+	var/datum/callback/expire_proc = null
+
+/datum/status_effect/delayed/on_creation(mob/living/new_owner, new_duration, datum/callback/new_expire_proc, new_prevent_signal = null)
+	if(!duration || !new_expire_proc || !istype(new_expire_proc))
```
Comment (SteelSlayer) [inline code/datums/status_effects/neutral.dm:230]: Think you meant `new_duration` here. We don't want this being null. Also `!istype(new_expire_proc)` doubles as a null check, so no need to do `!new_expire_proc`.
```suggestion
	if(!new_duration || !istype(new_expire_proc))
```

**Review by SteelSlayer — APPROVED** [review]

**Pull request merged**